New Logs Panel: Log line menu

matyax opened the pull request:
Part of https://github.com/grafana/grafana/issues/99075

This PR adds a log line menu to the new Logs Panel. To provide its required callbacks and improve the current situation 
where we rely on prop drilling to pass props to leaf components within the Logs visualization, a Log List Context component has been created, to allow the menu and any other component that requires props to read them from it, without having to modify and pass it through the tree.

Additionally, this PR adds unit tests for the new components.

Demo:

https://github.com/user-attachments/assets/4a8007d9-83b6-4d56-a1b9-205eedb0e894

matyax committed:
Create LogLineMenu component

matyax committed:
Fine tune icon width

matyax committed:
LogLineMenu: Add placeholder options

matyax committed:
utils: create reusable handleOpenLogsContextClick

matyax committed:
LogLineMenu: add callbacks to menu items

matyax committed:
LogListContext: create component

matyax committed:
LogList: use log list context to connect menu callbacks

matyax committed:
LogLine: add pinned style

matyax committed:
Remove unused imports

matyax committed:
LogLine: add unit test

matyax committed:
LogLine: add menu test case

matyax committed:
LogLineMenu: add unit test

matyax committed:
LogLineMessage: add unit test

matyax committed:
LogListContext: add unit test

matyax committed:
Remove unused code

matyax committed:
Extract translations

matyax commented on public/app/features/logs/components/LogRowMenuCell.tsx:
```diff
@@ -35,7 +36,6 @@ interface Props {
   styles: LogRowStyles;
   mouseIsOver: boolean;
   onBlur: () => void;
-  onPinToContentOutlineClick?: (row: LogRowModel, onOpenContext: (row: LogRowModel) => void) => void;
```
Not used.

matyax committed:
Fix handleOpenLogsContextClick

matyax committed:
Chore: memoize styles

matyax committed:
Virtualization: update node used for underflow detection

gtk-grafana commented on public/app/features/logs/components/panel/virtualization.ts:
```diff
@@ -6,8 +6,9 @@ import { LogListModel } from './processing';
 let gridSize = 8;
 let paddingBottom = gridSize * 0.75;
-let lineHeight = 22;
+export let lineHeight = 22;
```
Maybe split this up into a local lineHeight variable to be mutated and export a const defaultLogRowLineHeight variable?

matyax replied:
Yeah, good call.

gtk-grafana commented on public/app/features/logs/utils.ts:
```diff
@@ -303,6 +309,35 @@ export const copyText = async (text: string, buttonRef: React.MutableRefObject<E
+          range: getDefaultTimeRange(),
+        }),
+      });
+      win.location = url;
```
Take a look at locationService instead of using window

matyax replied:
This is copy paste from the current Panel, but definitely, let's update this.

matyax replied:
Just re-read the code and it's opening a new window with `const win = window.open('about:blank');`, so this doesn't look like a candidate for locationService. Let me know if I got this wrong.

gtk-grafana commented:
**Bugs:**
Menu doesn't close on selecting log context:
<img width="1728" alt="image" src="https://github.com/user-attachments/assets/3130d3dd-bf46-4177-a777-41aafea82a85" />

Log line doesn't pin with shared links (copy shortlink in old panel adds the selected line to pinned log lines automatically)

When pinning too many logs, the menu doesn't let you know you've hit the pinned limit.
<img width="1715" alt="image" src="https://github.com/user-attachments/assets/4bfd4424-d86d-442b-81e8-a07ee1487b85" />

Safari has two scrollbars:
https://github.com/user-attachments/assets/3e100a3e-9b9f-4545-94f5-1a832b034ec9


**Not anything to address in this PR:**
Would be nice if the pinned log lines could stick like they do in logs context
<img width="1633" alt="image" src="https://github.com/user-attachments/assets/3e493436-8092-41bd-8f53-3e1472aaba97" />

Line format doesn't seem to support ANSI codes like the old logs panel
<img width="1728" alt="image" src="https://github.com/user-attachments/assets/c4cc07d3-bcfa-4555-8d15-7df0a6c6cef6" />

Be nice if pinned log lines were shared in URL state.

matyax commented on public/app/features/logs/components/panel/InfiniteScroll.tsx:
```diff
@@ -60,7 +60,7 @@ export const InfiniteScroll = ({
   const lastLogOfPage = useRef<string[]>([]);
   const theme = useTheme2();
-  const styles = getStyles(theme);
+  const styles = useMemo(() => getStyles(theme), [theme]);
```
Oh definitely.

matyax replied:
Done in bbab5c9e0329223b27b2f9007842b7e3cfacf268

matyax committed:
Use useStyles2 instead of manually memoizing

matyax committed:
Virtualization: export getter instead of variable

matyax committed:
Open context: move stopPropagation to the old panel code

matyax committed:
Logs: add new container class

matyax commented:
> Menu doesn't close on selecting log context:

Great catch. Fixed in 4724ea614cd98d865881d24b5976446b666d58fc

> Log line doesn't pin with shared links (copy shortlink in old panel adds the selected line to pinned log lines automatically)

Does that really happen? Maybe I'm not understanding the comment, but I couldn't reproduce it.

> When pinning too many logs, the menu doesn't let you know you've hit the pinned limit.

The current one doesn't either. Created https://github.com/grafana/grafana/issues/101430

I would reimplement pinned logs by either actually pin them to the view, or pin them to both outline and viz. We can probably straight do the latter.

> Safari has two scrollbars:

Should be fixed by aa43876495c9bf0c6dd5075c90c9c2c4d35ff0a1

> Would be nice if the pinned log lines could stick like they do in logs context

Added to https://github.com/grafana/grafana/issues/99075 . I think we have an open issue to do that as a feature request.

> Line format doesn't seem to support ANSI codes like the old logs panel

Will address in the tokenization PR. https://github.com/grafana/grafana/pull/101401 I think if it has ansy, we should probably skip the tokenization. I'll think about it.

Great feedback, much appreciated.
